apphub/contrastive_learning/simclr/simclr_tf.py

```diff
@@ -0,0 +1,237 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""
+The FastEstimator implementation of SimCLR with ResNet9 on CIFAIR10.
+This code took reference from google implementation (https://github.com/google-research/simclr).
+Note that we use the ciFAIR10 dataset instead (https://cvjena.github.io/cifair/)
+"""
+import tempfile
+
+import tensorflow as tf
+from tensorflow.keras import layers
+
+import fastestimator as fe
+from fastestimator.dataset.data.cifair10 import load_data
+from fastestimator.op.numpyop.meta import Sometimes
+from fastestimator.op.numpyop.multivariate import HorizontalFlip, PadIfNeeded, RandomCrop
+from fastestimator.op.numpyop.univariate import ColorJitter, GaussianBlur, Onehot, ToFloat, ToGray
+from fastestimator.op.tensorop import LambdaOp, TensorOp
+from fastestimator.op.tensorop.loss import CrossEntropy
+from fastestimator.op.tensorop.model import ModelOp, UpdateOp
+from fastestimator.trace.io import BestModelSaver, ModelSaver
+from fastestimator.trace.metric import Accuracy
+
+
+def ResNet9(input_size=(32, 32, 3), head_len=128, classes=10):
+    """A small 9-layer ResNet Tensorflow model for cifar10 image classification.
+    The model architecture is from https://github.com/davidcpage/cifar10-fast
+
+    Args:
+        input_size: The size of the input tensor (height, width, channels).
+        classes: The number of outputs the model should generate.
+
+    Raises:
+        ValueError: Length of `input_size` is not 3.
+        ValueError: `input_size`[0] or `input_size`[1] is not a multiple of 16.
+
+    Returns:
+        A TensorFlow ResNet9 model.
+    """
+
+    # prep layers
+    inp = layers.Input(shape=input_size)
+    x = layers.Conv2D(64, 3, padding='same')(inp)
+    x = layers.BatchNormalization(momentum=0.8)(x)
+    x = layers.LeakyReLU(alpha=0.1)(x)
+    # layer1
+    x = layers.Conv2D(128, 3, padding='same')(x)
+    x = layers.MaxPool2D()(x)
+    x = layers.BatchNormalization(momentum=0.8)(x)
+    x = layers.LeakyReLU(alpha=0.1)(x)
+    x = layers.Add()([x, residual(x, 128)])
+    # layer2
+    x = layers.Conv2D(256, 3, padding='same')(x)
+    x = layers.MaxPool2D()(x)
+    x = layers.BatchNormalization(momentum=0.8)(x)
+    x = layers.LeakyReLU(alpha=0.1)(x)
+    # layer3
+    x = layers.Conv2D(512, 3, padding='same')(x)
+    x = layers.MaxPool2D()(x)
+    x = layers.BatchNormalization(momentum=0.8)(x)
+    x = layers.LeakyReLU(alpha=0.1)(x)
+    x = layers.Add()([x, residual(x, 512)])
+    # layers4
+    x = layers.GlobalMaxPool2D()(x)
+    code = layers.Flatten()(x)
+
+    p_head = layers.Dense(head_len)(code)
+    model_con = tf.keras.Model(inputs=inp, outputs=p_head)
```

Comment:
should we add an option to our fe resnet9 to return just the encoder block so that the model construction here can be simplified?